source/_posts/2018-01-12-pager-duty-and-confluence.markdown

```diff
@@ -0,0 +1,256 @@
+Viable Product ( MVP ) which allows us to prove out a business case.
+
+In this case we will prove out the value of incident reporting data.  This is a perfect example of a product
+that is valuable today, but becomes less valuable over time because it allows us to refine a process.
+
+We fully expect that some day this data will be effictively useless and thus this code will be thrown
+away.  However, the value of this code today is in it's ability to give us insights that we can act on
+today.
+
+We start this with a high-level HTTPClient.  Our nickname for this is called Groot: 
+Grand Rounds Operational and Organizational Tool.  We're all fans of the comic books over here, 
+so we often name our projects after Marvel characters.  I haven't seen anything from the DC 
+realm quite yet, but maybe someday.
+
+```ruby
+class GrootHTTPClient
+  @url
+  @headers
+
+  attr_accessor :url, :headers
+  def initialize
+    @url = get_url
+    @headers = get_headers
+  end
+
+  def get_headers
+    {
+        'Content-Type' => 'application/json'
+    }
+  end
+end
+```
+
+This is a very simply little base class to get us started.  Now let's build a specific client on top of this:
+
+```ruby
+class ConfluenceClient < GrootHTTPClient
+  @username
+  @password
+
+  attr_accessor :username, :password
+  def initialize
+    @username = ENV['JIRA_USERNAME'] ||= DEFAULT_USERNAME
+    @password = ENV['JIRA_PASSWORD'] ||= File.read(format('%s/.wiki_pass', ENV['HOME'])).chomp
+    super
+  end
+
+  def get_url
+    'https://HOSTED_SITE.atlassian.net/wiki'
+  end
+
+  def get_headers
+    super.merge({
+      Authorization: format('Basic %s', Base64.encode64(format('%s:%s', @username, @password)))
+    })
+  end
+
+  def get_page_by_id(page_id)
+    uri = format('%s/rest/api/content/%i?status=any', @url, page_id )
+    Log.debug(format('URI: %s', uri))
+    res = RestClient.get(uri, @headers)
+    JSON.parse res.body
+  end
+
+  def get_page(space_key, title)
+    params = { spaceKey: space_key, title: title, expand: 'version,body.view,space' }
+    param_str = params.map { |k,v| format('%s=%s', k, URI.encode(v)) }.join('&')
+    uri = format('%s/rest/api/content?%s', @url, param_str)
+    Log.debug(format('URI: %s', uri))
+    res = RestClient.get(uri, @headers)
+    JSON.parse( res.body )['results'].first
+  end
+
+  def update_page(page, new_content)
+    page_id = page['id'].to_i
+    current_version = page['version']['number']
+
+    params = { space: { key: page['space']['key'] }, title: page['title'], type: 'page' }
+    current_version += 1
+    params[:version] = {
+        hidden: false,
+        number: current_version,
+        minorEdit: true
+    }
+
+    params[:body] = {
+        storage: {
+            value: new_content,
+            representation: 'storage'
+        }
+    }
+
+    uri = format('%s/rest/api/content/%i', @url, page_id)
+    Log.debug(format('URI: %s', uri))
+    headers = @headers
```

Comment:
Are you doing this because you don't have an `attr_accessor` for `:headers`?

Reply:
I think I'm doing that because I've written this POS like 3 times over the lat few years so I'm pulling things from memory like a bat out of hell.